[nsan] Emit calls to optimized functions

chestnykh opened the pull request:
As previously noted in nsan.cpp we can implement
optimized variants of `__nsan_copy_values` and
`__nsan_set_value_unknown` if a memory operation
size is known.
Now the instrumentation creates calls to optimized functions if there is 4, 8 or 16-byte memory operation like
`memset(X, value, 4/8/16)` or `memcpy(dst, src, 4/8/16)`
nsan.cpp provides definitions of the optimized functions.

chestnykh committed:
[nsan] Emit calls to optimized functions

As previously noted in nsan.cpp we can implement
optimized variants of `__nsan_copy_values` and
`__nsan_set_value_unknown` if a memory operation
size is known.
Now the instrumentation creates calls to optimized functions
if there is 4, 8 or 16-byte memory operation like
`memset(X, value, 4/8/16)` or `memcpy(dst, src, 4/8/16)`
nsan.cpp provides definition of the optimized functions.

chestnykh commented:
CC: @MaskRay @alexander-shaposhnikov @vitalybuka

chestnykh committed:
[compiler-rt][nsan] Remove unneeded semicolons

chestnykh committed:
[nsan] Optimize `GetInstrumentationCalleeIdxForMemOp`

chestnykh committed:
[nsan] Simplify memory.ll test

- Test all calls to `__nsan_copy_*` in one function
- Test all calls to `__nsan_set_value_unknown_*` in one function

chestnykh commented:
CC: @nikic

alexander-shaposhnikov commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -2123,21 +2143,53 @@ bool NumericalStabilitySanitizer::sanitizeFunction(
 }
 
+static size_t GetInstrumentationCalleeIdxForMemOp(Value *V,
+                                                  std::size_t MaxIdx) {
```
for consistency one can drop std::

chestnykh replied:
Done

alexander-shaposhnikov commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -634,10 +637,27 @@ NumericalStabilitySanitizer::NumericalStabilitySanitizer(Module &M)
-                                         PtrTy, PtrTy, IntptrTy);
-  NsanSetValueUnknown = M.getOrInsertFunction("__nsan_set_value_unknown", Attr,
-                                              VoidTy, PtrTy, IntptrTy);
+  NsanCopyFunction[0] =
```
Since now we have multiple functions, arrays, and logic in GetInstrumentationCalleeIdxForMemOp, it'd be better to encapsulate them into helper classes, e.g., `class NsanCopy` and `class NsanSetUnknown`, to keep the internal details grouped together.

chestnykh replied:
Got it

chestnykh replied:
Done. I've tried to do it with one base class which implenents some basic virtual methods and a couple of derived class one per function family

chestnykh committed:
[nsan] Use size_t instead of std::size_t

chestnykh committed:
[nsan] Add helper classes

Add base class for handling details of
some `__nsan_*` functions
Add derived `NsanCopy` and `NsanSetUnknownClasses`
with the appropriate methods

chestnykh committed:
[nsan] Remove unneeded newline

chestnykh committed:
[nsan] Remove const qualifier from uint64_t args

alexander-shaposhnikov commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -493,6 +494,103 @@ class ValueToShadowMap {
 };
 
+// Base class for handling some details of __nsan_* functions
+class NsanInstrumentationFunction {
```
The classes share the interface, but they are not used polymorphically, so the base class doesn't do much  / I'd simplify things by removing it.

chestnykh replied:
Done in the latest commit

alexander-shaposhnikov commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -493,6 +494,103 @@ class ValueToShadowMap {
+  virtual ~NsanInstrumentationFunction() = default;
+};
+
+// Helper class for __nsan_copy_* family
```
How about simplifying / removing duplication, i.e. like this:
```

template <size_t N>
class MemOp {
public:
    MemOp(Module &M, ArrayRef<StringRef> Sized, StringRef Fallback);

    FunctionCallee getFunctionFor(MemIntrinsic &MI) const;    
    FunctionCallee getFallback() const;   
private:
    std::array<FunctionCallee, N> functions;
};

MemOp<4> NsanCopy;
MemOp<4> NsanSetUnknown;
```

chestnykh replied:
thanks for the tip, removing duplication is really something that would be nice here

chestnykh replied:
Done. I use the second template argument to pass information about number of arguments of the functions represented by the instance

chestnykh replied:
Or we can just pass this number as constructor argument?

chestnykh committed:
[nsan] Simplify helper classes

Use one class instead of base and derived classes

chestnykh committed:
[nsan] Pass number of fallback fn args as ctor param

chestnykh committed:
[nsan] Remove unneeded include

chestnykh committed:
[compiler-rt][nsan] Fix wrong formatting

clang-format cannot recognize that this is multiplication,
not pointer declaration.
See #99271

chestnykh commented on compiler-rt/lib/nsan/nsan.cpp:
```diff
@@ -64,13 +62,33 @@ __nsan_copy_values(const u8 *daddr, const u8 *saddr, uptr size) {
+    __builtin_memmove((void *)GetShadowTypeAddrFor(daddr),                     \
+                      GetShadowTypeAddrFor(saddr), N);                         \
+    __builtin_memmove((void *)GetShadowAddrFor(daddr),                         \
+                      GetShadowAddrFor(saddr), N *kShadowScale);               \
```
Here clang-format failed to format the code correctly.
`N *kShadowScale` means 'pointer of type N' when we have multiplication here.
https://github.com/llvm/llvm-project/issues/99271

chestnykh commented:
I've force-pushed because bots couldn't start jobs for this PR
p.s. code formatting fails because of the clang-format issue

alexander-shaposhnikov commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -493,6 +493,58 @@ class ValueToShadowMap {
+
+  for (size_t i = 0; i < N - 1; ++i) {
+    if (NFallbackArgs == 3)
+      Funcs[i] = M.getOrInsertFunction(Sized[i], Attr, VoidTy, PtrTy, PtrTy);
```
this can be simplified further - getOrInsertFunction has other overloads that take function type (can be computed once based on the number of parameters)

chestnykh replied:
Done, thx

alexander-shaposhnikov commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -493,6 +493,58 @@ class ValueToShadowMap {
+
+template <size_t N>
+NsanMemOpFn<N>::NsanMemOpFn(Module &M, ArrayRef<StringRef> Sized,
+                            StringRef Fallback, int NFallbackArgs) {
```
nit: int NFallBackArgs -> size_t NumArgs

chestnykh replied:
Done

alexander-shaposhnikov commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -598,7 +654,14 @@ static GlobalValue *createThreadLocalGV(const char *Name, Module &M, Type *Ty) {
+      NsanCopyFns(M, {"__nsan_copy_4", "__nsan_copy_8", "__nsan_copy_16"},
+                  "__nsan_copy_values", 3),
+      NsanSetUnknownFns(M,
+                        {"__nsan_set_value_unknown_4",
```
nit: `git clang-format --style=llvm`

chestnykh replied:
this command tells me that there are no modified files and the style is correct

chestnykh replied:
I've corrected codestyle of previous commits, but these lines stay unchanged

chestnykh committed:
[nsan] Calculate function type once

With this approach the constructor of `NsanMemOpFn`
can be simplified

chestnykh commented:
Sorry, i've pushed wrong commits

chestnykh committed:
[nsan] Update comment

chestnykh committed:
[nsan] Correct code style

alexander-shaposhnikov commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -493,6 +493,59 @@ class ValueToShadowMap {
 };
 
+// Template parameter is the number of functions
+template <size_t N> class NsanMemOpFn {
```
After taking a closer look - maybe push it even further and drop the template parameter ?
The number of sized variants is known in the constructor (Sized.size()),
std::array can be replaced with SmallVector (don't think the size of the data structure is a particular concern here).
p.s. If the code in getFunctionFor is specific to a particular value of N, we should add a comment and an assert (or asserts).

chestnykh replied:
Yes, good idea. Template parameter is not really needed anymore here

chestnykh replied:
Done

alexander-shaposhnikov commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -598,7 +655,14 @@ static GlobalValue *createThreadLocalGV(const char *Name, Module &M, Type *Ty) {
-    : DL(M.getDataLayout()), Context(M.getContext()), Config(Context) {
+    : DL(M.getDataLayout()), Context(M.getContext()), Config(Context),
+      NsanCopyFns(M, {"__nsan_copy_4", "__nsan_copy_8", "__nsan_copy_16"},
+                  "__nsan_copy_values", 3),
```
1. Why 3 and 2 (2 and 1 ?) ?
2. Need a comment  (e.g. 2 /*NumArgs*/)

chestnykh replied:
Done

alexander-shaposhnikov commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -493,6 +493,59 @@ class ValueToShadowMap {
+  for (size_t i = 0; i < N - 1; ++i)
+    Funcs[i] = M.getOrInsertFunction(Sized[i], SizedFnTy, Attr);
+
+  if (NumArgs == 3)
```
nit: it'd look a bit more natural to have a single if-else:
```
if (NumArgs == 3) {
    Funcs.back() =
        M.getOrInsertFunction(Fallback, Attr, VoidTy, PtrTy, PtrTy, IntptrTy);
    SizedFnTy = FunctionType::get(VoidTy, {PtrTy, PtrTy}, false);
} else if (NumArgs == 2) {
    Funcs.back() =
        M.getOrInsertFunction(Fallback, Attr, VoidTy, PtrTy,  IntptrTy);
    SizedFnTy = FunctionType::get(VoidTy, {PtrTy}, false);
} else {
    assert(!"Unexpected value of NumArgs");
}
```

chestnykh replied:
Dont, thx

chestnykh committed:
[nsan] Remove template parameter and refactor code

chestnykh committed:
[nsan] Remove wrong comments

alexander-shaposhnikov commented:
To me looks good.
P.S. Do you mind if we hold off on landing this for a couple of days - I wanted to publish a PR with integration tests, this would give us a bit more confidence that things still work. In the meantime, would be great to try to build a ~relatively large program with this PR, although you might encounter other issues (most likely https://github.com/llvm/llvm-project/issues/98143)

alexander-shaposhnikov commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -493,6 +493,67 @@ class ValueToShadowMap {
+public:
+  NsanMemOpFn(Module &M, ArrayRef<StringRef> Sized, StringRef Fallback,
+              size_t NumArgs);
+  // Number of parameters can be extracted from FunctionCallee
```
nit: this comment can be dropped.

chestnykh replied:
Done

chestnykh committed:
[nsan] Reorganize work with SmallVector

chestnykh committed:
[nsan] Drop unneeded comment

chestnykh commented:
@alexander-shaposhnikov i've pushed two commits a couple of minutes ago, look please one more time, especially on operations with SmallVector of Funcs. Are they correct?

chestnykh commented:
> To me looks good. P.S. Do you mind if we hold off on landing this for a couple of days - I wanted to publish a PR with integration tests, this would give us a bit more confidence that things still work. In the meantime, would be great to try to build a ~relatively large program with this PR, although you might encounter other issues (most likely #98143)

Of course, no problem. Validation of such a change with tests is great

chestnykh commented:
> To me looks good. P.S. Do you mind if we hold off on landing this for a couple of days - I wanted to publish a PR with integration tests, this would give us a bit more confidence that things still work. In the meantime, would be great to try to build a ~relatively large program with this PR, although you might encounter other issues (most likely #98143)

ping me please when integration tests land, i will need to rebase my PR

chestnykh commented:
> To me looks good. P.S. Do you mind if we hold off on landing this for a couple of days - I wanted to publish a PR with integration tests, this would give us a bit more confidence that things still work. In the meantime, would be great to try to build a ~relatively large program with this PR, although you might encounter other issues (most likely #98143)

I used eigen math library and samples from [here](https://yang.amp.i.kyoto-u.ac.jp/~yyama/computer/FAQ/eigen/eigensystem.html) .

In assembler output i see calls to some new functions like __nsan_set_value_unknown_16, __nsan_copy_* (they differ at different optimizations levels). And the programs work correctly

chestnykh commented:
Rebase to check this PR on the nsan tests that was landed recently

chestnykh commented:
Force pushed to remove unrelated changes added accidentally

alexander-shaposhnikov commented:
@chestnykh - do you want me to try to merge this PR ?

chestnykh commented:
> @chestnykh - do you want me to try to merge this PR ?

Yes,you can merge it, but lets wait for checks to complete

chestnykh commented:
Checked locally that `make check-nsan` suceeds

hctim commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -493,6 +493,60 @@ class ValueToShadowMap {
+        M.getOrInsertFunction(Fallback, Attr, VoidTy, PtrTy, IntptrTy));
+    SizedFnTy = FunctionType::get(VoidTy, {PtrTy}, false);
+  } else {
+    assert(!"Unexpected value of sized functions arguments");
```
Looks like this breaks under buildbots with `-Werror`: https://lab.llvm.org/buildbot/#/builders/109/builds/1300

```
[3128/3995] Building CXX object lib/Transforms/Instrumentation/CMakeFiles/LLVMInstrumentation.dir/NumericalStabilitySanitizer.cpp.o
FAILED: lib/Transforms/Instrumentation/CMakeFiles/LLVMInstrumentation.dir/NumericalStabilitySanitizer.cpp.o 
CCACHE_CPP2=yes CCACHE_HASHDIR=yes /usr/bin/ccache /home/buildbots/llvm-external-buildbots/workers/ppc64be-sanitizer/sanitizer-ppc64be/build/llvm_build0/bin/clang++ -DGTEST_HAS_RTTI=0 -D_DEBUG -D_GLIBCXX_ASSERTIONS -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/home/buildbots/llvm-external-buildbots/workers/ppc64be-sanitizer/sanitizer-ppc64be/build/build_default/lib/Transforms/Instrumentation -I/home/buildbots/llvm-external-buildbots/workers/ppc64be-sanitizer/sanitizer-ppc64be/build/llvm-project/llvm/lib/Transforms/Instrumentation -I/home/buildbots/llvm-external-buildbots/workers/ppc64be-sanitizer/sanitizer-ppc64be/build/build_default/include -I/home/buildbots/llvm-external-buildbots/workers/ppc64be-sanitizer/sanitizer-ppc64be/build/llvm-project/llvm/include -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wc++98-compat-extra-semi -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -O3 -DNDEBUG -std=c++17  -fno-exceptions -funwind-tables -fno-rtti -UNDEBUG -MD -MT lib/Transforms/Instrumentation/CMakeFiles/LLVMInstrumentation.dir/NumericalStabilitySanitizer.cpp.o -MF lib/Transforms/Instrumentation/CMakeFiles/LLVMInstrumentation.dir/NumericalStabilitySanitizer.cpp.o.d -o lib/Transforms/Instrumentation/CMakeFiles/LLVMInstrumentation.dir/NumericalStabilitySanitizer.cpp.o -c /home/buildbots/llvm-external-buildbots/workers/ppc64be-sanitizer/sanitizer-ppc64be/build/llvm-project/llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp
/home/buildbots/llvm-external-buildbots/workers/ppc64be-sanitizer/sanitizer-ppc64be/build/llvm-project/llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:530:13: error: implicit conversion turns string literal into bool: 'const char[46]' to 'bool' [-Werror,-Wstring-conversion]
  530 |     assert(!"Unexpected value of sized functions arguments");
      |            ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/assert.h:87:5: note: expanded from macro 'assert'
   87 |   ((expr)                                                               \
      |     ^~~~
1 error generated.
```
